dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java

```diff
@@ -293,9 +470,965 @@ public void resetContextCache() {
     }
 
     void setContextualProperty(String key, Object v) {
-        if (!containsKey(key)) {
-            put(key, v);
+        putIfAbsent(key, v);
+    }
+    
+    @SuppressWarnings("rawtypes")
+    public Map getProtocolHeaders() {
+        return (Map) getFromPropertyArray(protoHeaders);
+    }
+    
+    @SuppressWarnings("rawtypes")
+    public void setProtocolHeaders(Map p) {
+        propertyValues[protoHeaders] = p;
+    }
+    
+    @Override
+    public Object remove(Object key) {
+        return remove((String) key);
+    }
+    
+    public Object remove(String key) {
+        Integer index = KEYMAP.getOrDefault(key, KEY_NOT_FOUND);
+        if (index != KEY_NOT_FOUND) {
+            Object ret = getFromPropertyArray(index);
+            propertyValues[index] = NOT_FOUND;
+            return ret;
+        }
+        return super.remove(key);
+    }
+    
+    @Override
+    public Object get(Object key) {
+        return get((String) key);
+    }
+
+    public Object get(String key) {
+        Integer index = KEYMAP.getOrDefault(key, KEY_NOT_FOUND);
+        if (index != KEY_NOT_FOUND) {
+            return getFromPropertyArray(index);
+        }
+        
+        return super.get(key);
+    }
+
+    @Override
+    public Object put(String key, Object value) {
+        Integer index = KEYMAP.getOrDefault(key, KEY_NOT_FOUND);
+        if (index != KEY_NOT_FOUND) {
+            Object ret = getFromPropertyArray(index);
+            propertyValues[index] = value;
+            return ret;
+        }
+
+        return super.put(key, value);
+    }
+
+    @Override
+    public Set<String> keySet() {
+        return keySet != null ? keySet : (keySet = new KeySet());
+    }
+    
+    abstract class MessageIterator<T> implements Iterator<T>{
+        Iterator<T> backedIterator;
+        int current = -1;
+        int next = -1;
+        boolean removeAllowed = false;
+        MessageIterator(Iterator<T> it) {
+            backedIterator = it;
+        }
+        @Override
+        public final boolean hasNext() {
+            if (backedIterator.hasNext()) {
+                return true;
+            }
+            if (next == TOTAL) {
+                return false;
+            }
+
```

Comment:
This should have a conditional in it that says
```
if (next > current) {
    return true;
}
```

This handles the hasNext(), hasNext() being called back to back without calling next.  No reason to calculate next again if we already did it.